tensorboard/tools/whitespace_hygiene_test.py

```diff
@@ -0,0 +1,135 @@
+    "tensorboard/plugins/mesh/http_api.md",
+    "tensorboard/plugins/mesh/mesh_plugin_test.py",
+    "tensorboard/plugins/mesh/metadata_test.py",
+    "tensorboard/plugins/mesh/summary_test.py",
+    "tensorboard/plugins/mesh/test_utils.py",
+    "tensorboard/plugins/pr_curve/http_api.md",
+    "tensorboard/plugins/profile/README.md",
+    "tensorboard/plugins/profile/pod_viewer/stack_bar_chart/stack-bar-chart.ts",
+    "tensorboard/plugins/projector/vz_projector/umap.d.ts",
+    "tensorboard/plugins/projector/vz_projector/util.ts",
+    "tensorboard/plugins/projector/vz_projector/vz-projector-inspector-panel.ts",
+    "tensorboard/plugins/projector/vz_projector/vz-projector-metadata-card.html",
+    "tensorboard/plugins/projector/vz_projector/vz-projector-projections-panel.html",
+    "tensorboard/plugins/projector/vz_projector/vz-projector-projections-panel.ts",
+    "tensorboard/plugins/projector/vz_projector/vz-projector.ts",
+    "tensorboard/program_test.py",
+    "tensorboard/tools/docs_list_format_test.sh",
+])
+
+
+Match = collections.namedtuple("Match", ("filename", "line_number", "line"))
+
+
+def main():
+  chdir_to_repo_root()
+  matches = git_grep("  *$")
+  errors = [m for m in matches if m.filename not in exceptions]
+  okay = True
+
+  if errors:
+    print("Superfluous trailing whitespace:")
+    for error in errors:
+      print("%s:%d:%s$" % (error.filename, error.line_number, error.line))
+    print()
+    okay = False
+
+  stale_exceptions = exceptions - frozenset(m.filename for m in matches)
+  if stale_exceptions:
+    print("Stale exceptions (no whitespace problems; prune exceptions list):")
+    for filename in stale_exceptions:
+      print(filename)
+    print()
+    okay = False
+
+  sys.exit(0 if okay else 1)
+
+
+def git_grep(pattern):
+  """Run `git grep` and collect matches.
+
+  This function exits the process if `git grep` writes any stderr: for
+  instance, if the provided pattern is an invalid regular expression.
+
+  Args:
+    pattern: `str`; a pattern argument to `git grep`.
+
+  Returns:
+    A list of `Match` values.
+  """
+  cmd = ["git", "grep", "-Izn", "--", pattern]
+  p = subprocess.Popen(cmd, stdout=subprocess.PIPE, stderr=subprocess.PIPE)
+  (stdout, stderr) = p.communicate()
+  if stderr:
+    getattr(sys.stderr, "buffer", sys.stderr).write(stderr)  # Python 2 compat
+    sys.exit(1)
+  result = []
+  while stdout:
+    (filename_raw, _unused_nul, rest) = stdout.partition(b"\0")
+    (line_number_raw, _unused_nul, rest) = rest.partition(b"\0")
+    (line_raw, _unused_lf, stdout) = rest.partition(b"\n")
+    match = Match(
+        filename=filename_raw.decode("utf-8", errors="replace"),
+        line_number=int(line_number_raw),
+        line=line_raw.decode("utf-8", errors="replace"),
+    )
+    result.append(match)
+  return result
+
+
+def chdir_to_repo_root():
+  toplevel = subprocess.check_output(["git", "rev-parse", "--show-toplevel"])
+  toplevel = toplevel[:-1]  # trim trailing LF
```

Comment:
`os.chdir(toplevel.rstrip())`?

Reply:
This is again for whitespace-correctness—nothing wrong with

```
$ git clone -q gh:tensorflow/tensorboard "tensorboard "
$ cd "tensorboard "
$ pwd | xxd
00000000: 2f74 6d70 2f74 6d70 2e49 426d 7057 5446  /tmp/tmp.IBmpWTF
00000010: 7573 712f 7465 6e73 6f72 626f 6172 6420  usq/tensorboard 
00000020: 0a                                       .
$ python -c '
> import os
> import subprocess
> toplevel = subprocess.check_output(["git", "rev-parse", "--show-toplevel"])
> os.chdir(toplevel.rstrip())
> '
Traceback (most recent call last):
  File "<string>", line 5, in <module>
OSError: [Errno 2] No such file or directory: '/tmp/tmp.IBmpWTFusq/tensorboard'
```

It really costs us nothing here to do the right thing, and I’d love to
avoid proliferating even more programs that behave incorrectly on
filenames that have whitespace.